### PR TITLE
Issue287/no activity ref in application

### PR DESCRIPTION
**Pull request by maiatoday**
Assorted memory leak fixes:
The biggest one was removing the reference to an activity from the application. Multiple leak canary leaks were appearing because of this.  I reran all of the UseCase tests and the PopupNotificationManager tests and they passed after the change.

Also I made the RemoteConfig not be a singleton and rather replaced with just method calls. From the existing code a singleton isn't needed. I left some comments in there because I don't understand the history or intention of this so maybe we want it back.

I moved the eventbus register and unregister into the basefragment so we know it is always called.

This does not close Issue 287. There are still some more leaks I am seeing with Leak Canary but I it is looking much better.

I tested by putting breakpoints in the code I changed and going to that part of the app manually. This does not mean I haven't broken anything. We probably need to do a thorough test.

my Habitica User-ID: c34cbae3-a66f-4211-884f-0c90d48d71fc

### CODE REVIEW

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java [hunk 1]
```diff
@@ -133,5 +108,32 @@ public String getDisplayedClassName() {
         return this.getClass().getSimpleName();
     }
 
+    private class TutorialStepTransactionListener implements TransactionListener<TutorialStep> {
```
Comment (vIiRuS) [inline Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java]: this can be a static inner class, right?

File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/RemoteConfigManager.java [hunk 2]
```diff
@@ -28,29 +28,27 @@
...
-            instance = new RemoteConfigManager(context);
-        }
-        return instance;
-    }
+//    public static RemoteConfigManager getInstance(Context context) {
+//        if (instance == null) {
+//            instance = new RemoteConfigManager(context);
+//        }
+//        return instance;
+//    }
```
Comment (vIiRuS) [inline Habitica/src/main/java/com/habitrpg/android/habitica/helpers/RemoteConfigManager.java]: if this isn't needed anymore it should be removed instead of commented out.

File: Habitica/build.gradle [hunk 3]
```diff
@@ -306,6 +306,5 @@ tasks.whenTaskAdded { task ->
 
 check { findbugs { skip true } }
 
-apply plugin: 'com.android.application' //or apply plugin: 'java'
```
Comment (vIiRuS) [inline Habitica/build.gradle]: why was this removed? Isn't it needed?

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java (hunk 1 above)
Comment (maiatoday) [inline Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java]: it's the static inner classes that keep a reference to the outer class if they remain after the activity goes away

File: Habitica/build.gradle (hunk 3 above)
Comment (maiatoday) [inline Habitica/build.gradle]: it was duplicated

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java (hunk 1 above)
Comment (vIiRuS) [inline Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java]: I think it's the other way around? non-static inner classes have a reference to the outer class, but static ones don't.

Comment (maiatoday) [inline Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java]: sorry misread, anonymous inner classes make the leak. Let me revisit the code.

File: Habitica/src/main/java/com/habitrpg/android/habitica/helpers/RemoteConfigManager.java (hunk 2 above)
Comment (maiatoday) [inline Habitica/src/main/java/com/habitrpg/android/habitica/helpers/RemoteConfigManager.java]: ok I'll remove

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java (hunk 1 above)
Comment (maiatoday) [inline Habitica/src/main/java/com/habitrpg/android/habitica/ui/fragments/BaseFragment.java]: Static innerclasses don't keep a reference to the object, they belong to the class. Making TutorialStepTransactionListener a static inner class will remove the reference however it then requires tutorialText, tutorialTexts and tutorialCanBeDeferred to be static variables as they are needed inside the static innerclass. This means these three variables are then shared among all fragments that derive from BaseFragment and they also become GC roots.The original version (non-static inner class) made the leak canary fault go away. I will change to a static inner and retest with leak canary and check again. Will update with more info when I have it.

**Comment by maiatoday** [comment]
I am going to close this pull request. I will redo the changes on the latest develop branch as many things have changed. I will make a new pull request when that has been done.
